Sources/StreamChat/StateLayer/CurrentUserState.swift

```diff
@@ -0,0 +1,36 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+@available(iOS 13.0, *)
+@dynamicMemberLookup
+public final class CurrentUserState: ObservableObject {
+    private let observer: Observer
+    
+    init(user: CurrentChatUser, database: DatabaseContainer) {
+        self.observer = Observer(database: database)
+        self.user = user
+        observer.start(
+            with: .init(
+                userDidChange: { [weak self] user in await self?.setValue(user, for: \.user) })
+        )
+    }
+    
+    /// The represented user.
+    @Published public private(set) var user: CurrentChatUser
+    
+    // MARK: - Dynamic Member Lookup
+    
+    /// Provides data about the represented user.
+    public subscript<T>(dynamicMember keyPath: KeyPath<CurrentChatUser, T>) -> T {
+        user[keyPath: keyPath]
+    }
+    
+    // MARK: - Mutating the State
+    
+    @MainActor func setValue<Value>(_ value: Value, for keyPath: ReferenceWritableKeyPath<CurrentUserState, Value>) {
+        self[keyPath: keyPath] = value
+    }
```

Comment:
I'm not sure I fully understood why do we need this 🤔 Why can't we simple update the state without the keypaths?

Reply:
It is used for forcing mutations on the main thread because it is an ObservableObject and therefore highly likely it is used by SDK consumers directly in SwiftUI views. In that case we do not want to cause runtime issues for them that something is triggering SwiftUI view update from a background thread.

When I was adding this, I also considered:

**Option 1**
Make `XYZState: ObservableObject` to require `@MainActor`: `@MainActor XYZState: ObservableObject`
Pro: Everything is main actor, therefore mutation is always happen on the main actor
Con: Harder to use this state type since existing code needs to adopt MainActor as well (our SwiftUI view models do not use `@MainActor`)

**Option 2**
Add a convenience @MainActor forced key-path based function for mutating stuff. 
Pro: Easy to adopt the state object since its public interface does not require @MainActor isolation
Con: It is not guarded against data corruption. Think about background threads reading public properties and at the same time main thread mutating the state. 

To be fair, we never had a real discussion between these options. Do you happen to have a preference on this matter? Other State objects currently follow the same pattern. If we decide to change it, it should be done for all the `XYZState` objects. I am happy to discuss this.

Reply:
Slack discussion was concluded to not expose main actor on the public interface level. Therefore, let's just use it for internal implementation and forcing mutation on the main thread.